samples/otlptrace/example_grpc.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python3
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+import google.auth
+import google.auth.transport.grpc
+import google.auth.transport.requests
+import grpc
+from google.auth.transport.grpc import AuthMetadataPlugin
+from opentelemetry import trace
+from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import (
+    OTLPSpanExporter,
+)
+from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import BatchSpanProcessor
+
+"""
+This is a sample script that exports OTLP traces encoded as protobufs via gRPC. 
+"""
+
+
+class AutoRefreshAuthMetadataPlugin(AuthMetadataPlugin):
+    """
+    A `gRPC AuthMetadataPlugin`_ that refreshes credentials and inserts them into
+    each request.
+    """
+
+    def __init__(self, credentials, request, default_host=None):
+        super().__init__(credentials, request, default_host)
+
+    def __call__(self, context, callback):
+        if self._credentials.expired:
+            self._credentials.refresh(self._request)
+
+        auth_headers = [("authorization", f"Bearer {self._credentials.token}")]
+        callback(auth_headers, None)
+
+
+credentials, project_id = google.auth.default()
+request = google.auth.transport.requests.Request()
+credentials.refresh(request)
+resource = Resource.create(attributes={SERVICE_NAME: "otlp-gcp-grpc-sample"})
+
+auth_metadata_plugin = AutoRefreshAuthMetadataPlugin(
```

Comment:
(Continuing the other thread here)

I was able to get it working with `AuthMetadataPlugin` directly and it ran successfully overnight.
```diff
diff --git a/samples/otlptrace/example_grpc.py b/samples/otlptrace/example_grpc.py                                                           
index cb53cd1..f8d1820 100644
--- a/samples/otlptrace/example_grpc.py                                                                                                      
+++ b/samples/otlptrace/example_grpc.py
@@ -55,7 +55,7 @@ request = google.auth.transport.requests.Request()                                                                         
 credentials.refresh(request)
 resource = Resource.create(attributes={SERVICE_NAME: "otlp-gcp-grpc-sample"})
 
-auth_metadata_plugin = AutoRefreshAuthMetadataPlugin(
+auth_metadata_plugin = AuthMetadataPlugin(                                                                                                  
     credentials=credentials, request=request
 )
 channel_creds = grpc.composite_channel_credentials(
@@ -68,9 +68,6 @@ processor = BatchSpanProcessor(
     OTLPSpanExporter(
         credentials=channel_creds,
         insecure=False,
-        headers={
-            "x-goog-user-project": credentials.quota_project_id,
-        },
     )
 )
 trace_provider.add_span_processor(processor)
```

The problem was L72 adds `x-goog-user-project` but this header was also being added by the AuthMetadataPlugin causing duplicate headers to be comma joined and an `INVALID_ARGUMENT` response.

Reply:
Good find! Thanks @aabmass !